File: packages/core/src/mcpact/contract/engine.py
"""Run a contract's assertions against a connected MCP client."""

from __future__ import annotations

import re
from typing import Any

from jsonschema import Draft202012Validator
from jsonschema.exceptions import ValidationError as JsonSchemaError

from ..client.base import CallOutcome, McpClient, ResourceInfo, ToolInfo
from ..report import CheckResult, CheckStatus, Report
from .hooks import run_hooks
from .models import (
    Assertion,
    Contract,
    Expectation,
    InputSchemaAssertion,
    PromptSpec,
    ResourceSpec,
    ToolSpec,
)
from .variables import VariableError, resolve_value


async def run_contract(
    contract: Contract,
    client: McpClient,
    *,
    fail_fast: bool = False,
    no_hooks: bool = False,
) -> Report:
    """Execute every check the contract declares and return a Report.

    When `fail_fast` is True, stop as soon as any check fails.
    When `no_hooks` is True, before/after hooks are skipped.
    """
    report = Report()

    if not no_hooks and contract.before:
        try:
            await run_hooks(contract.before, client)
        except Exception as e:
            report.add(CheckResult("contract", "before", CheckStatus.FAIL, f"hook error: {e}"))
            return report

    try:
        tools = await client.list_tools()
        by_name = {t.name: t for t in tools}

        for spec in contract.tools:
            tool = by_name.get(spec.name)
            if tool is None:
                if spec.must_exist:
                    report.add(
                        CheckResult(spec.name, "exists", CheckStatus.FAIL, "tool not exposed by server")
                    )
                    if fail_fast:
                        return report
                else:
                    report.add(CheckResult(spec.name, "exists", CheckStatus.SKIP, "optional tool absent"))
                continue
            report.add(CheckResult(spec.name, "exists", CheckStatus.PASS))

            _check_description(spec, tool, report)
            _check_input_schema(spec, tool, report)
            if fail_fast and report.failed:
                return report

            if not no_hooks and spec.before:
                try:
                    await run_hooks(spec.before, client)
                except Exception as e:
                    report.add(CheckResult(spec.name, "before", CheckStatus.FAIL, f"hook error: {e}"))
                    continue

            step_context: dict[str, Any] = {}
            try:
                for idx, assertion in enumerate(spec.assertions, start=1):
                    await _run_assertion(spec.name, idx, assertion, client, report, step_context)
                    if fail_fast and report.failed:
                        return report
            finally:
                if not no_hooks and spec.after:
                    try:
                        await run_hooks(spec.after, client, step_context)
                    except Exception as e:
                        report.add(CheckResult(spec.name, "after", CheckStatus.FAIL, f"hook error: {e}"))

        if contract.resources:
            await _run_resources(contract.resources, client, report, fail_fast)
            if fail_fast and report.failed:
                return report

        if contract.prompts:
            await _run_prompts(contract.prompts, client, report, fail_fast)
    finally:
        if not no_hooks and contract.after:
            try:
                await run_hooks(contract.after, client)
            except Exception as e:
                report.add(CheckResult("contract", "after", CheckStatus.FAIL, f"hook error: {e}"))

    return report


# ---------- resources ----------

async def _run_resources(
    specs: list[ResourceSpec],
    client: McpClient,
    report: Report,
    fail_fast: bool,
) -> None:
    resources = await client.list_resources()
    for spec in specs:
        if spec.uri:
            await _check_resource_by_uri(spec, resources, client, report)
        else:
            _check_resource_by_pattern(spec, resources, report)
        if fail_fast and report.failed:
            return


async def _check_resource_by_uri(
    spec: ResourceSpec,
    resources: list[ResourceInfo],
    client: McpClient,
    report: Report,
) -> None:
    assert spec.uri is not None
    subject = f"resource:{spec.uri}"
    found = any(r.uri == spec.uri for r in resources)
    if not found:
        if spec.must_exist:
            report.add(CheckResult(subject, "exists", CheckStatus.FAIL, "resource not listed"))
        else:
            report.add(CheckResult(subject, "exists", CheckStatus.SKIP, "optional resource absent"))
        return
    report.add(CheckResult(subject, "exists", CheckStatus.PASS))

    if spec.content_contains is not None or spec.content_schema is not None:
        try:
            content = await client.read_resource(spec.uri)
        except Exception as e:
            report.add(
                CheckResult(subject, "read", CheckStatus.FAIL, f"read failed: {e}")
            )
            return
        _check_resource_content(subject, spec, content.text, report)


def _check_resource_by_pattern(
    spec: ResourceSpec,
    resources: list[ResourceInfo],
    report: Report,
) -> None:
    assert spec.uri_pattern is not None
    subject = f"resource:~{spec.uri_pattern}"
    try:
        pattern = re.compile(spec.uri_pattern)
    except re.error as e:
        report.add(CheckResult(subject, "pattern", CheckStatus.FAIL, f"invalid regex: {e}"))
        return
    matched = [r for r in resources if pattern.search(r.uri)]
    if spec.min_count is not None and len(matched) < spec.min_count:
        report.add(
            CheckResult(
                subject,
                "min_count",
                CheckStatus.FAIL,
                f"matched {len(matched)}, expected >= {spec.min_count}",
            )
        )
    elif spec.min_count is not None:
        report.add(CheckResult(subject, "min_count", CheckStatus.PASS))
    if not matched and spec.must_exist:
        report.add(CheckResult(subject, "exists", CheckStatus.FAIL, "no resources match pattern"))


def _check_resource_content(
    subject: str, spec: ResourceSpec, text: str, report: Report
) -> None:
    if spec.content_contains is not None:
        needles = (
            [spec.content_contains]
            if isinstance(spec.content_contains, str)
            else list(spec.content_contains)
        )
        missing = [n for n in needles if n not in text]
        if missing:
            report.add(
                CheckResult(subject, "content_contains", CheckStatus.FAIL, f"missing: {missing!r}")
            )
        else:
            report.add(CheckResult(subject, "content_contains", CheckStatus.PASS))

    if spec.content_schema is not None:
        import json

        try:
            payload = json.loads(text)
        except json.JSONDecodeError:
            report.add(
                CheckResult(subject, "content_schema", CheckStatus.FAIL, "content is not valid JSON")
            )
            return
        try:
            Draft202012Validator(spec.content_schema).validate(payload)
        except JsonSchemaError as e:
            report.add(
                CheckResult(
                    subject,
                    "content_schema",
                    CheckStatus.FAIL,
                    f"schema mismatch: {e.message}",
                )
            )
            return
        report.add(CheckResult(subject, "content_schema", CheckStatus.PASS))


# ---------- prompts ----------

async def _run_prompts(
    specs: list[PromptSpec],
    client: McpClient,
    report: Report,
    fail_fast: bool,
) -> None:
    prompts = await client.list_prompts()
    by_name = {p.name: p for p in prompts}
    for spec in specs:
        prompt = by_name.get(spec.name)
        subject = f"prompt:{spec.name}"
        if prompt is None:
            if spec.must_exist:
                report.add(CheckResult(subject, "exists", CheckStatus.FAIL, "prompt not listed"))
            else:
                report.add(CheckResult(subject, "exists", CheckStatus.SKIP))
            continue
        report.add(CheckResult(subject, "exists", CheckStatus.PASS))

        if spec.description_contains is not None:
            needles = (
                [spec.description_contains]
                if isinstance(spec.description_contains, str)
                else list(spec.description_contains)
            )
            desc = prompt.description or ""
            missing = [n for n in needles if n not in desc]
            if missing:
                report.add(
                    CheckResult(
                        subject, "description_contains", CheckStatus.FAIL,
                        f"description missing: {missing!r}",
                    )
                )
            else:
                report.add(CheckResult(subject, "description_contains", CheckStatus.PASS))

        if spec.arguments:
            prompt_arg_names = {a.name for a in prompt.arguments}
            for arg_spec in spec.arguments:
                check = f"arg:{arg_spec.name}"
                if arg_spec.name not in prompt_arg_names:
                    report.add(CheckResult(subject, check, CheckStatus.FAIL, "argument not declared"))
                else:
                    report.add(CheckResult(subject, check, CheckStatus.PASS))
                    if arg_spec.required:
                        actual = next(a for a in prompt.arguments if a.name == arg_spec.name)
                        if not actual.required:
                            report.add(
                                CheckResult(subject, f"{check}.required", CheckStatus.FAIL, "expected required")
                            )
                        else:
                            report.add(CheckResult(subject, f"{check}.required", CheckStatus.PASS))

        for pa_idx, pa in enumerate(spec.assertions, start=1):
            await _run_prompt_assertion(subject, pa_idx, spec.name, pa, client, report)

        if fail_fast and report.failed:
            return


async def _run_prompt_assertion(
    subject: str,
    idx: int,
    prompt_name: str,
    pa: Any,
    client: McpClient,
    report: Report,
) -> None:
    from .models import PromptAssertion

    assert isinstance(pa, PromptAssertion)
    label = f"get#{idx}"
    try:
        result = await client.get_prompt(prompt_name, dict(pa.get_prompt.args))
    except Exception as e:
        report.add(
            CheckResult(subject, label, CheckStatus.FAIL, f"get_prompt error: {e}")
        )
        return

    expect = pa.expect
    if expect.message_count is not None:
        actual = len(result.messages)
        if actual != expect.message_count:
            report.add(
                CheckResult(
                    subject, f"{label}.message_count", CheckStatus.FAIL,
                    f"expected {expect.message_count}, got {actual}",
                )
            )
        else:
            report.add(CheckResult(subject, f"{label}.message_count", CheckStatus.PASS))

    if expect.messages_contain is not None:
        needles = (
            [expect.messages_contain]
            if isinstance(expect.messages_contain, str)
            else list(expect.messages_contain)
        )
        all_text = " ".join(m.text for m in result.messages)
        missing = [n for n in needles if n not in all_text]
        if missing:
            report.add(
                CheckResult(
                    subject, f"{label}.messages_contain", CheckStatus.FAIL,
                    f"missing: {missing!r}",
                )
            )
        else:
            report.add(CheckResult(subject, f"{label}.messages_contain", CheckStatus.PASS))

    if expect.messages_schema is not None:
        payload = [{"role": m.role, "text": m.text} for m in result.messages]
        try:
            Draft202012Validator(expect.messages_schema).validate(payload)
        except JsonSchemaError as e:
            report.add(
                CheckResult(
                    subject, f"{label}.messages_schema", CheckStatus.FAIL,
                    f"schema mismatch: {e.message}",
                )
            )
        else:
            report.add(CheckResult(subject, f"{label}.messages_schema", CheckStatus.PASS))


# ---------- tools ----------

def _check_description(spec: ToolSpec, tool: ToolInfo, report: Report) -> None:
    if spec.description_contains is None:
        return
    needles = (
        [spec.description_contains]
        if isinstance(spec.description_contains, str)
        else list(spec.description_contains)
    )
    desc = tool.description or ""
    missing = [n for n in needles if n not in desc]
    if missing:
        report.add(
            CheckResult(
                spec.name,
                "description_contains",
                CheckStatus.FAIL,
                f"description missing: {missing!r}",
            )
        )
    else:
        report.add(CheckResult(spec.name, "description_contains", CheckStatus.PASS))


def _check_input_schema(spec: ToolSpec, tool: ToolInfo, report: Report) -> None:
    if spec.input_schema is None:
        return
    assertion: InputSchemaAssertion = spec.input_schema
    schema = tool.input_schema or {}
    required = set(schema.get("required", []) or [])
    properties = schema.get("properties", {}) or {}

    missing_required = [r for r in assertion.required if r not in required]
    if missing_required:
        report.add(
            CheckResult(
                spec.name,
                "input_schema.required",
                CheckStatus.FAIL,
                f"missing required fields: {missing_required}",
            )
        )
    elif assertion.required:
        report.add(CheckResult(spec.name, "input_schema.required", CheckStatus.PASS))

    for prop_name, expected in assertion.properties.items():
        actual = properties.get(prop_name)
        if actual is None:
            report.add(
                CheckResult(
                    spec.name,
                    f"input_schema.properties.{prop_name}",
                    CheckStatus.FAIL,
                    "property not declared by tool",
                )
            )
            continue
        expected_type = expected.get("type")
        if expected_type and actual.get("type") != expected_type:
            report.add(
                CheckResult(
                    spec.name,
                    f"input_schema.properties.{prop_name}",
                    CheckStatus.FAIL,
                    f"expected type {expected_type!r}, got {actual.get('type')!r}",
                )
            )
        else:
            report.add(
                CheckResult(
                    spec.name, f"input_schema.properties.{prop_name}", CheckStatus.PASS
                )
            )


async def _run_assertion(
    tool_name: str,
    idx: int,
    assertion: Assertion,
    client: McpClient,
    report: Report,
    step_context: dict[str, Any] | None = None,
) -> None:
    label = assertion.name or f"call#{idx}"
    ctx = step_context if step_context is not None else {}

    try:
        resolved_args = resolve_value(dict(assertion.call.args), ctx)
    except VariableError as e:
        report.add(
            CheckResult(tool_name, label, CheckStatus.FAIL, f"variable error: {e}")
        )
        return

    try:
        outcome = await client.call_tool(
            tool_name, resolved_args, timeout_ms=assertion.call.timeout_ms
        )
    except Exception as e:
        report.add(
            CheckResult(
                tool_name,
                label,
                CheckStatus.FAIL,
                f"transport error: {type(e).__name__}: {e}",
            )
        )
        return

    if assertion.name:
        ctx[assertion.name] = {
            "text": outcome.text,
            "structured": outcome.structured,
            "is_error": outcome.is_error,
        }

    _check_status(tool_name, label, assertion.expect, outcome, report)
    _check_latency(tool_name, label, assertion.expect, outcome, report)
    _check_response_contains(tool_name, label, assertion.expect, outcome, report)
    _check_response_schema(tool_name, label, assertion.expect, outcome, report)


def _check_status(
    tool: str, label: str, expect: Expectation, outcome: CallOutcome, report: Report
) -> None:
    expected_error = expect.status == "error"
    actual_error = outcome.is_error
    if expected_error == actual_error:
        report.add(
            CheckResult(
                tool,
                f"{label}.status",
                CheckStatus.PASS,
                latency_ms=outcome.latency_ms,
            )
        )
    else:
        msg = f"expected status={expect.status!r}, got {'error' if actual_error else 'success'}"
        if outcome.text:
            msg += f" — {outcome.text[:200]}"
        report.add(
            CheckResult(
                tool,
                f"{label}.status",
                CheckStatus.FAIL,
                msg,
                latency_ms=outcome.latency_ms,
            )
        )


def _check_latency(
    tool: str, label: str, expect: Expectation, outcome: CallOutcome, report: Report
) -> None:
    if expect.max_latency_ms is None:
        return
    if outcome.latency_ms <= expect.max_latency_ms:
        report.add(
            CheckResult(
                tool,
                f"{label}.max_latency_ms",
                CheckStatus.PASS,
                latency_ms=outcome.latency_ms,
            )
        )
    else:
        report.add(
            CheckResult(
                tool,
                f"{label}.max_latency_ms",
                CheckStatus.FAIL,
                f"{outcome.latency_ms:.0f}ms > {expect.max_latency_ms}ms",
                latency_ms=outcome.latency_ms,
            )
        )


def _check_response_contains(
    tool: str, label: str, expect: Expectation, outcome: CallOutcome, report: Report
) -> None:
    if expect.response_contains is None:
        return
    needles = (
        [expect.response_contains]
        if isinstance(expect.response_contains, str)
        else list(expect.response_contains)
    )
    missing = [n for n in needles if n not in outcome.text]
    if missing:
        report.add(
            CheckResult(
                tool,
                f"{label}.response_contains",
                CheckStatus.FAIL,
                f"missing substrings: {missing!r}",
            )
        )
    else:
        report.add(CheckResult(tool, f"{label}.response_contains", CheckStatus.PASS))


def _check_response_schema(
    tool: str, label: str, expect: Expectation, outcome: CallOutcome, report: Report
) -> None:
    if expect.schema_ is None:
        return
    payload: Any = outcome.structured if outcome.structured is not None else outcome.text
    try:
        Draft202012Validator(expect.schema_).validate(payload)
    except JsonSchemaError as e:
        report.add(
            CheckResult(
                tool,
                f"{label}.schema",
                CheckStatus.FAIL,
                f"schema mismatch at {'/'.join(map(str, e.absolute_path))}: {e.message}",
            )
        )
        return
    report.add(CheckResult(tool, f"{label}.schema", CheckStatus.PASS))

File: packages/sdk-node/src/contract/engine.ts
import Ajv from "ajv";
import type { McpClient } from "../client/base.js";
import type { CallOutcome, ResourceInfo, ToolInfo } from "../types.js";
import { Report } from "../types.js";
import type {
  Assertion,
  Contract,
  Expectation,
  PromptAssertion,
  PromptSpec,
  ResourceSpec,
  ToolSpec,
} from "./schema.js";
import { VariableError, resolveValue } from "./variables.js";

const ajv = new Ajv({ strict: false, allErrors: true });

export async function runContract(contract: Contract, client: McpClient): Promise<Report> {
  const report = new Report();
  const tools = await client.listTools();
  const byName = new Map(tools.map((t) => [t.name, t]));

  for (const spec of contract.tools) {
    const tool = byName.get(spec.name);
    if (!tool) {
      if (spec.must_exist) {
        report.add({
          subject: spec.name,
          check: "exists",
          status: "fail",
          message: "tool not exposed by server",
        });
      } else {
        report.add({
          subject: spec.name,
          check: "exists",
          status: "skip",
          message: "optional tool absent",
        });
      }
      continue;
    }
    report.add({ subject: spec.name, check: "exists", status: "pass" });

    checkDescription(spec, tool, report);
    checkInputSchema(spec, tool, report);

    const stepContext: Record<string, unknown> = {};
    let i = 0;
    for (const assertion of spec.assertions) {
      i += 1;
      await runAssertion(spec.name, i, assertion, client, report, stepContext);
    }
  }

  if (contract.resources.length > 0) {
    await runResources(contract.resources, client, report);
  }
  if (contract.prompts.length > 0) {
    await runPrompts(contract.prompts, client, report);
  }

  return report;
}

// ---------- resources ----------

async function runResources(
  specs: ResourceSpec[],
  client: McpClient,
  report: Report,
): Promise<void> {
  const resources = await client.listResources();
  for (const spec of specs) {
    if (spec.uri !== undefined) {
      await checkResourceByUri(spec, resources, client, report);
    } else if (spec.uri_pattern !== undefined) {
      checkResourceByPattern(spec, resources, report);
    }
  }
}

async function checkResourceByUri(
  spec: ResourceSpec,
  resources: ResourceInfo[],
  client: McpClient,
  report: Report,
): Promise<void> {
  const uri = spec.uri!;
  const subject = `resource:${uri}`;
  const found = resources.some((r) => r.uri === uri);
  if (!found) {
    if (spec.must_exist) {
      report.add({ subject: subject, check: "exists", status: "fail", message: "resource not listed" });
    } else {
      report.add({ subject: subject, check: "exists", status: "skip", message: "optional resource absent" });
    }
    return;
  }
  report.add({ subject: subject, check: "exists", status: "pass" });

  if (spec.content_contains !== undefined || spec.content_schema !== undefined) {
    let text: string;
    try {
      const content = await client.readResource(uri);
      text = content.text;
    } catch (e) {
      report.add({ subject: subject, check: "read", status: "fail", message: `read failed: ${e instanceof Error ? e.message : String(e)}` });
      return;
    }
    checkResourceContent(subject, spec, text, report);
  }
}

function checkResourceByPattern(
  spec: ResourceSpec,
  resources: ResourceInfo[],
  report: Report,
): void {
  const pat = spec.uri_pattern!;
  const subject = `resource:~${pat}`;
  let re: RegExp;
  try {
    re = new RegExp(pat);
  } catch (e) {
    report.add({ subject: subject, check: "pattern", status: "fail", message: `invalid regex: ${e instanceof Error ? e.message : String(e)}` });
    return;
  }
  const matched = resources.filter((r) => re.test(r.uri));
  if (spec.min_count !== undefined) {
    if (matched.length < spec.min_count) {
      report.add({ subject: subject, check: "min_count", status: "fail", message: `matched ${matched.length}, expected >= ${spec.min_count}` });
    } else {
      report.add({ subject: subject, check: "min_count", status: "pass" });
    }
  }
  if (matched.length === 0 && spec.must_exist) {
    report.add({ subject: subject, check: "exists", status: "fail", message: "no resources match pattern" });
  }
}

function checkResourceContent(subject: string, spec: ResourceSpec, text: string, report: Report): void {
  if (spec.content_contains !== undefined) {
    const needles = typeof spec.content_contains === "string" ? [spec.content_contains] : spec.content_contains;
    const missing = needles.filter((n) => !text.includes(n));
    if (missing.length > 0) {
      report.add({ subject: subject, check: "content_contains", status: "fail", message: `missing: ${JSON.stringify(missing)}` });
    } else {
      report.add({ subject: subject, check: "content_contains", status: "pass" });
    }
  }
  if (spec.content_schema !== undefined) {
    let payload: unknown;
    try {
      payload = JSON.parse(text);
    } catch {
      report.add({ subject: subject, check: "content_schema", status: "fail", message: "content is not valid JSON" });
      return;
    }
    const validate = ajv.compile(spec.content_schema);
    if (validate(payload)) {
      report.add({ subject: subject, check: "content_schema", status: "pass" });
    } else {
      const first = (validate.errors ?? [])[0];
      report.add({ subject: subject, check: "content_schema", status: "fail", message: `schema mismatch: ${first?.message ?? "unknown"}` });
    }
  }
}

// ---------- prompts ----------

async function runPrompts(
  specs: PromptSpec[],
  client: McpClient,
  report: Report,
): Promise<void> {
  const prompts = await client.listPrompts();
  const byName = new Map(prompts.map((p) => [p.name, p]));

  for (const spec of specs) {
    const subject = `prompt:${spec.name}`;
    const prompt = byName.get(spec.name);
    if (!prompt) {
      if (spec.must_exist) {
        report.add({ subject: subject, check: "exists", status: "fail", message: "prompt not listed" });
      } else {
        report.add({ subject: subject, check: "exists", status: "skip" });
      }
      continue;
    }
    report.add({ subject: subject, check: "exists", status: "pass" });

    if (spec.description_contains !== undefined) {
      const needles = typeof spec.description_contains === "string" ? [spec.description_contains] : spec.description_contains;
      const desc = prompt.description ?? "";
      const missing = needles.filter((n) => !desc.includes(n));
      if (missing.length > 0) {
        report.add({ subject: subject, check: "description_contains", status: "fail", message: `description missing: ${JSON.stringify(missing)}` });
      } else {
        report.add({ subject: subject, check: "description_contains", status: "pass" });
      }
    }

    if (spec.arguments.length > 0) {
      const names = new Set(prompt.arguments.map((a) => a.name));
      for (const as of spec.arguments) {
        const check = `arg:${as.name}`;
        if (!names.has(as.name)) {
          report.add({ subject: subject, check, status: "fail", message: "argument not declared" });
        } else {
          report.add({ subject: subject, check, status: "pass" });
          if (as.required) {
            const actual = prompt.arguments.find((a) => a.name === as.name);
            if (!actual?.required) {
              report.add({ subject: subject, check: `${check}.required`, status: "fail", message: "expected required" });
            } else {
              report.add({ subject: subject, check: `${check}.required`, status: "pass" });
            }
          }
        }
      }
    }

    let pi = 0;
    for (const pa of spec.assertions) {
      pi += 1;
      await runPromptAssertion(subject, pi, spec.name, pa, client, report);
    }
  }
}

async function runPromptAssertion(
  subject: string,
  idx: number,
  promptName: string,
  pa: PromptAssertion,
  client: McpClient,
  report: Report,
): Promise<void> {
  const label = `get#${idx}`;
  let result: Awaited<ReturnType<McpClient["getPrompt"]>>;
  try {
    result = await client.getPrompt(promptName, pa.get_prompt.args);
  } catch (e) {
    report.add({ subject: subject, check: label, status: "fail", message: `get_prompt error: ${e instanceof Error ? e.message : String(e)}` });
    return;
  }

  const expect = pa.expect;
  if (expect.message_count !== undefined) {
    if (result.messages.length !== expect.message_count) {
      report.add({ subject: subject, check: `${label}.message_count`, status: "fail", message: `expected ${expect.message_count}, got ${result.messages.length}` });
    } else {
      report.add({ subject: subject, check: `${label}.message_count`, status: "pass" });
    }
  }

  if (expect.messages_contain !== undefined) {
    const needles = typeof expect.messages_contain === "string" ? [expect.messages_contain] : expect.messages_contain;
    const allText = result.messages.map((m) => m.text).join(" ");
    const missing = needles.filter((n) => !allText.includes(n));
    if (missing.length > 0) {
      report.add({ subject: subject, check: `${label}.messages_contain`, status: "fail", message: `missing: ${JSON.stringify(missing)}` });
    } else {
      report.add({ subject: subject, check: `${label}.messages_contain`, status: "pass" });
    }
  }

  if (expect.messages_schema !== undefined) {
    const payload = result.messages.map((m) => ({ role: m.role, text: m.text }));
    const validate = ajv.compile(expect.messages_schema);
    if (validate(payload)) {
      report.add({ subject: subject, check: `${label}.messages_schema`, status: "pass" });
    } else {
      const first = (validate.errors ?? [])[0];
      report.add({ subject: subject, check: `${label}.messages_schema`, status: "fail", message: `schema mismatch: ${first?.message ?? "unknown"}` });
    }
  }
}

function asList(v: string | string[] | undefined): string[] {
  if (v === undefined) return [];
  return typeof v === "string" ? [v] : v;
}

function checkDescription(spec: ToolSpec, tool: ToolInfo, report: Report): void {
  const needles = asList(spec.description_contains);
  if (needles.length === 0) return;
  const desc = tool.description ?? "";
  const missing = needles.filter((n) => !desc.includes(n));
  if (missing.length > 0) {
    report.add({
      subject: spec.name,
      check: "description_contains",
      status: "fail",
      message: `description missing: ${JSON.stringify(missing)}`,
    });
  } else {
    report.add({ subject: spec.name, check: "description_contains", status: "pass" });
  }
}

function checkInputSchema(spec: ToolSpec, tool: ToolInfo, report: Report): void {
  const want = spec.input_schema;
  if (!want) return;
  const schema = tool.inputSchema ?? {};
  const required = new Set(Array.isArray(schema["required"]) ? (schema["required"] as string[]) : []);
  const properties = (schema["properties"] as Record<string, { type?: string }> | undefined) ?? {};

  const missingRequired = want.required.filter((r) => !required.has(r));
  if (missingRequired.length > 0) {
    report.add({
      subject: spec.name,
      check: "input_schema.required",
      status: "fail",
      message: `missing required fields: ${JSON.stringify(missingRequired)}`,
    });
  } else if (want.required.length > 0) {
    report.add({ subject: spec.name, check: "input_schema.required", status: "pass" });
  }

  for (const [propName, expected] of Object.entries(want.properties)) {
    const actual = properties[propName];
    const check = `input_schema.properties.${propName}`;
    if (!actual) {
      report.add({
        subject: spec.name,
        check,
        status: "fail",
        message: "property not declared by tool",
      });
      continue;
    }
    const expectedType = (expected as { type?: string }).type;
    if (expectedType && actual.type !== expectedType) {
      report.add({
        subject: spec.name,
        check,
        status: "fail",
        message: `expected type '${expectedType}', got '${actual.type ?? "undefined"}'`,
      });
    } else {
      report.add({ subject: spec.name, check, status: "pass" });
    }
  }
}

async function runAssertion(
  toolName: string,
  idx: number,
  assertion: Assertion,
  client: McpClient,
  report: Report,
  stepContext: Record<string, unknown> = {},
): Promise<void> {
  const label = assertion.name ?? `call#${idx}`;

  let resolvedArgs: Record<string, unknown>;
  try {
    resolvedArgs = resolveValue(assertion.call.args, stepContext) as Record<string, unknown>;
  } catch (e) {
    if (e instanceof VariableError) {
      report.add({ subject: toolName, check: label, status: "fail", message: `variable error: ${e.message}` });
      return;
    }
    throw e;
  }

  let outcome: CallOutcome;
  try {
    outcome = await client.callTool(
      toolName,
      resolvedArgs,
      assertion.call.timeout_ms !== undefined ? { timeoutMs: assertion.call.timeout_ms } : {},
    );
  } catch (e) {
    report.add({
      subject: toolName,
      check: label,
      status: "fail",
      message: `transport error: ${e instanceof Error ? e.message : String(e)}`,
    });
    return;
  }

  if (assertion.name !== undefined) {
    stepContext[assertion.name] = {
      text: outcome.text,
      structured: outcome.structured,
      is_error: outcome.isError,
    };
  }

  checkStatus(toolName, label, assertion.expect, outcome, report);
  checkLatency(toolName, label, assertion.expect, outcome, report);
  checkResponseContains(toolName, label, assertion.expect, outcome, report);
  checkResponseSchema(toolName, label, assertion.expect, outcome, report);
}

function checkStatus(
  subject: string,
  label: string,
  expect: Expectation,
  outcome: CallOutcome,
  report: Report,
): void {
  const expectedError = expect.status === "error";
  const check = `${label}.status`;
  if (expectedError === outcome.isError) {
    report.add({ subject, check, status: "pass", latencyMs: outcome.latencyMs });
    return;
  }
  let message = `expected status='${expect.status}', got '${outcome.isError ? "error" : "success"}'`;
  if (outcome.text) message += ` — ${outcome.text.slice(0, 200)}`;
  report.add({ subject, check, status: "fail", message, latencyMs: outcome.latencyMs });
}

function checkLatency(
  subject: string,
  label: string,
  expect: Expectation,
  outcome: CallOutcome,
  report: Report,
): void {
  if (expect.max_latency_ms === undefined) return;
  const check = `${label}.max_latency_ms`;
  if (outcome.latencyMs <= expect.max_latency_ms) {
    report.add({ subject, check, status: "pass", latencyMs: outcome.latencyMs });
  } else {
    report.add({
      subject,
      check,
      status: "fail",
      message: `${Math.round(outcome.latencyMs)}ms > ${expect.max_latency_ms}ms`,
      latencyMs: outcome.latencyMs,
    });
  }
}

function checkResponseContains(
  subject: string,
  label: string,
  expect: Expectation,
  outcome: CallOutcome,
  report: Report,
): void {
  const needles = asList(expect.response_contains);
  if (needles.length === 0) return;
  const check = `${label}.response_contains`;
  const missing = needles.filter((n) => !outcome.text.includes(n));
  if (missing.length > 0) {
    report.add({
      subject,
      check,
      status: "fail",
      message: `missing substrings: ${JSON.stringify(missing)}`,
    });
  } else {
    report.add({ subject, check, status: "pass" });
  }
}

function checkResponseSchema(
  subject: string,
  label: string,
  expect: Expectation,
  outcome: CallOutcome,
  report: Report,
): void {
  if (!expect.schema) return;
  const check = `${label}.schema`;
  const payload: unknown = outcome.structured ?? outcome.text;
  const validate = ajv.compile(expect.schema);
  if (validate(payload)) {
    report.add({ subject, check, status: "pass" });
    return;
  }
  const first = (validate.errors ?? [])[0];
  const at = first?.instancePath || "/";
  const msg = first?.message ?? "schema mismatch";
  report.add({ subject, check, status: "fail", message: `schema mismatch at ${at}: ${msg}` });
}
